Ipvs: remove module check

uablrek opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

The check of kernel modules in proxy/ipvs prevents usage with statically linked kernels. Instead the *function* should be checked.


#### Which issue(s) this PR fixes:

Fixes #108579

#### Special notes for your reviewer:

Please read https://github.com/kubernetes/kubernetes/issues/108579#issuecomment-1125675626.

Comments on individual commits below.


#### Does this PR introduce a user-facing change?

```release-note
kube-proxy with proxy-mode=ipvs can be used with statically linked kernels.
The reseved IPv4 range TEST-NET-2 in rfc5737 MUST NOT be used for ClusterIP or loadBalancerIP since address 198.51.100.0 is used for probing.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

Commit:
proxy/ipvs: Remove kernel module tests

To check kernel modules is a bad way to check functionality.
This commit just removes the checks and makes it possible to
use a statically linked kernel.

Minimal updates to unit-tests are made.

Commit:
proxy/ipvs: Check that a dummy virtual server can be added

This tests both ipvs and the configured scheduler

uablrek commented:
### Commit - proxy/ipvs: Remove kernel module tests 

Just removes the module checks

uablrek commented:
### proxy/ipvs: Check that a dummy virtual server can be added

The logic should have been straight forward if not for the BUG:
```
	// BUG: If ipvs is not compiled into the kernel ipvs.GetVirtualServers
	// does not return an error. Instead also ipvs.AddVirtualServer returns OK
	// (no error)!
```
As a work-around I re-do ipvs.GetVirtualServers after adding the dummy server and expect a len > 0. I will investigate this bug further. I hope it is in K8s, but it can be in "github.com/moby/ipvs".

I build my own kernel so I can test this. I have built one kernel without `ipvs` entirely, and one with ipvs but without the `lc` scheduler. Below are some log examples.

Succesful start:
```
I1222 20:09:40.117013     400 proxier.go:735] "Virtual Servers" count=0
I1222 20:09:40.119420     400 proxier.go:765] "Virtual Servers after adding dummy" count=1
I1222 20:09:40.119451     400 proxier.go:771] "Dummy VS created" vs={Address:10.0.0.1 Protocol:TCP Port:20000 Scheduler:rr Flags:0 Timeout:0}
I1222 20:09:40.121132     400 server_others.go:248] "Using ipvs Proxier"
```
With ipvs, but without lc and lc configured as sheduler:
```
I1222 20:11:58.764112     391 proxier.go:735] "Virtual Servers" count=0
E1222 20:11:58.765903     391 proxier.go:755] "Could not create dummy VS" err="no such file or directory" scheduler="lc"
E1222 20:11:58.765930     391 server.go:492] "Error running ProxyServer" err="can't use the IPVS proxier: no such file or directory"
```
Kernel without ipvs:
```
I1222 20:13:44.631763     399 proxier.go:735] "Virtual Servers" count=0
I1222 20:13:44.631834     399 proxier.go:765] "Virtual Servers after adding dummy" count=0
I1222 20:13:44.631847     399 proxier.go:767] "Dummy VS not created" scheduler="rr"
E1222 20:13:44.631868     399 server.go:492] "Error running ProxyServer" err="can't use the IPVS proxier: Ipvs not supported"
```
This shows the bug. No error is returned from GetVirtualServers, nor AddVirtualServer even though there is no ipvs at all.

uablrek commented:
/assign @thockin

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -720,51 +720,58 @@ func (handle *LinuxKernelHandler) GetKernelVersion() (string, error) {
+	// This is the first time kube-proxy is loaded on the node. Try to
+	// insert a dummy VS with the passed scheduler. The VIP address can be
+	// anything since no traffic will be routed to ipvs yet.
+	vs := utilipvs.VirtualServer{
```
I don't know about this - there's not really any magic IP or port that we can use which is guaranteed NOT to collide with existing services.  Is there a different protocol we can use or some other way to be SURE that it isn't really used?  And what if it exists but was a leftover from a previous crash?  Maybe you should force-delete first, ignoring errors, then create (assuming we keep this mechanism).

uablrek replied:
The ipvs is empty when this is executed (no virtual servers). To get traffic to it the address must be assigned to interface `kube-ipvs0`, so it _should_ be safe.

But I can surely find a more obscure test address. For instance an ipv6 address reserver in [fc3849](https://www.rfc-editor.org/rfc/rfc3849) like 2001:DB8::3333 which should *never* be used.

thockin replied:
That would help (though beware nodes without IPv6?) - a great big comment here about why this is "safe" would be nice.

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -720,51 +720,58 @@ func (handle *LinuxKernelHandler) GetKernelVersion() (string, error) {
-	if err != nil {
-		return fmt.Errorf("error determining kernel version to find required kernel modules for ipvs support: %v", err)
+	klog.V(5).InfoS("Virtual Servers", "count", len(vservers))
+	if len(vservers) > 0 {
```
This does not confirm the scheduler works - is that OK?  Why do we confirm that later and not here?

uablrek replied:
My reasoning is that if virtual servers are configured in ipvs already, then kube-proxy has been succesfully added them earlier, so it is re-started, possibly upgraded, but no check of ipvs and the scheduler is needed.

That assumes that kube-proxy is the only used of ipvs, and that a change of the configured scheduled is a forbidden operation (at least not without a node reboot).

thockin replied:
What if the user changes the kube-proxy config to a new scheduler - is it fair to demand a reboot?  Will they know to do that?

uablrek replied:
> What if the user changes the kube-proxy config to a new scheduler - is it fair to demand a reboot? Will they know to do that?

Good point. I have made an unnecessary assumption here. I will cover that and test.

Commit:
Updates after rewiew

To update the scheduler without node reboot now works.
The address for the probe VS is now 198.51.100.0 which is
reseved for documentation, please see rfc5737. The comment
about this is extended.

Commit:
gofmt

uablrek commented:
### Updates after rewiew

Address "198.51.100.0" is now used for the probe VS. It is reserved for documentation only by [rfc5737](https://datatracker.ietf.org/doc/html/rfc5737). To use that address IRL is an invalid configuration. The release-note is updated in this PR. The comment is extended:

```
	// We should use a VIP address that is not used on the node.
	// An address "198.51.100.0" from the TEST-NET-2 rage in https://datatracker.ietf.org/doc/html/rfc5737
	// is used. These addresses are reserved for documentation. If the user is using
	// this address for a VS anyway we *will* mess up, but that would be an invalid configuration.
	// If the user have configured the address to an interface on the node (but not a VS)
	// then traffic will temporary be routed to ipvs during the probe and dropped.
	// The later case is also and invalid configuration, but the traffic impact will be minor.
	// This should not be a problem if users honors reserved addresses, but cut/paste
	// from documentation is not unheard of, so the restricion to not use the TEST-NET-2 range
	// must be documented.
```

I have tested to re-configure the scheduler without node-reboot and it works fine.

uablrek commented:
#### About automatic kernel module loading

The kernel request a module by an up-call to:
```
> cat /proc/sys/kernel/modprobe
/sbin/modprobe
```
This is used for instance by iptables to load `xt_*` modules like xt_MASQUERADE.

It is however possible to break this and enforce manual module loading. Example:
```
> echo /bin/false > /proc/sys/kernel/modprobe
> ipvsadm -A -s mh -t 10.2.2.2:8080
Scheduler or persistence engine not found
> modprobe ip_vs_mh
> ipvsadm -A -s mh -t 10.2.2.2:8080
# (now works)
```

Commit:
Clean-up un-used code

Commit:
Fixed typo

Commit:
Remove obsolete test case

uablrek commented:
Slack thread in sig-docs on the test address https://kubernetes.slack.com/archives/C1J0BPD2M/p1671793688732989

Commit:
gofmt

Commit:
Handle an empty scheduler ("")

Commit:
use netutils.ParseIPSloppy

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -720,61 +651,88 @@ func (handle *LinuxKernelHandler) GetKernelVersion() (string, error) {
-
-	kernelVersionStr, err := handle.GetKernelVersion()
+// The ipset version and the scheduler are checked. If any virtual servers (VS)
+// already exist with the configured scheduler, we just return. Otherwise
```
Ohh, fun corner case.  Suppose a user was using IPVS mode with a valid scheduler, and has many VSes.  They then change it to an invalid scheduler and restart kube-proxy.  We detect that they have VSes, but none with this new scheduler.  We try the dummy and it fails.  Should we:

a) Throw an error and abort / panic / exit - existing VSes will work, nothing new will work

b) Fall back on a safe scheduler -  existing VSes will work, new VSes will work but maybe not optimally, users may never see the fact that it failed

c) Fall back on iptables - existing VSes will be abandoned, possibly conflicting with iptables logic, users may never see the fact that it failed

d) (b) or (c) plus events indicating a serious failure, users may STILL never see those unless they go looking

uablrek replied:
For now it is a) if an error is returned from CanUseIPVSProxier(), kube-proxy terminates.

It feels like the right thing to do in CanUseIPVSProxier() is to return an error (as now) and let the caller decide.

It is not simple (I thingk) to fallback to the working scheduler in the CanUseIPVSProxier() function since that mean to alter the global kube-proxy config from a low-level function. So my suggestion is to let it be as-is.

uablrek replied:
I thought kube-proxy would fall-back to iptables, but when I test it terminates. Log messages;
```
E1223 18:48:05.252466     401 server.go:492] "Error running ProxyServer" err="content of the proxy server's configuration file was updated"
E1223 18:48:05.252606     401 run.go:74] "command failed" err="content of the proxy server's configuration file was updated"
(restart...)
I1223 18:48:07.351241     830 proxier.go:685] "Virtual Servers" count=1
I1223 18:48:07.351253     830 proxier.go:687] "Virtual server exists" count=1
I1223 18:48:07.351260     830 proxier.go:695] "No existing VS uses the configured scheduler" scheduler="lc"
E1223 18:48:07.352400     830 proxier.go:716] "Could not create dummy VS" err="no such file or directory" scheduler="lc"
E1223 18:48:07.352419     830 server.go:492] "Error running ProxyServer" err="can't use the IPVS proxier: no such file or directory"
E1223 18:48:07.352432     830 run.go:74] "command failed" err="can't use the IPVS proxier: no such file or directory"
(restart forever...)
```
(the one VS is the kubernetes svc)

thockin replied:
Yeah, we removed the fallback logic in 1609017f2b5911356656e9edabeac3fb44fca1e5

thockin replied:
I agree (a) is the best choice

uablrek commented:
Unfortunately the ipvs detect bug was in moby: https://github.com/moby/ipvs/issues/27

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -720,61 +651,88 @@ func (handle *LinuxKernelHandler) GetKernelVersion() (string, error) {
+		Port:      20000,
+		Scheduler: scheduler,
+	}
+	if err := ipvs.AddVirtualServer(&vs); err != nil {
```
What happens if we do this, then crash, then restart?  When we get back to this function, we'll find this VS exists (the above cases) and it may or may not have the correct scheduler.  If it does, we will never delete it (again in the `len(vservers) > 0` path above).

Perhaps a little extra paranoia about cleaning this magic entry up?

uablrek replied:
There is no difference from today, the CanUseIPVSProxier() returns OK. But the kube-proxy seems to handle it. If I change the sceduler the kube-proxy re-starts. Whatever sheduler is finds if re-configures if CanUseIPVSProxier() retrurns ok. That is not part of this PR that just alters the method of determining if ipvs and the scheduler can be used, but it seem to work.

Before;
```
# ipvsadm -Ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  12.0.0.1:443 rr
  -> 192.168.1.1:6443             Masq    1      0          0         
```
then the scheduler is updated to "wlc", the kube-config detects the change and re-starts.
After;
```
# ipvsadm -Ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  12.0.0.1:443 wlc
  -> 192.168.1.1:6443             Masq    1      0          0         
```
Frankly, I don't know how it does, probably a complete re-sync, but it seem to be very solid.

uablrek replied:
No old VS'es seem ever to linger. I can only test basic cases, but the return from CanUseIPVSProxier() is not altered, only the method of getting there.

uablrek replied:
BTW, I tested to skip the removal of the probe-VS just to see it with "ipvsadm", but it was cleared away. I suspect that a "Flush()" is called somewhere, but I havn't investigated.

thockin replied:
I think we want to know why.  If I put a panic() between L730 and L731 (just before the delete), I would expect that it would restart and take the above happy-path, retruning at L686.

If not, we should document why, don't you think?

uablrek replied:
Merry Christmas. My boy has a new desktop gaming PC and is unrechable from planet earth, so I got time for a test. I describe it in the main conversation area.

thockin replied:
Thanks for the explanation in the main conversation.

thockin commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -622,73 +620,6 @@ func NewLinuxKernelHandler() *LinuxKernelHandler {
-			bmods = append(bmods, module)
-		} else {
-			// Try to load the required IPVS kernel modules if not built in
-			err := handle.executor.Command("modprobe", "--", module).Run()
```
So we are sure we don't need to modprobe?  We shouldn't need to, but since we were doing it, I just want to confirm :)

uablrek replied:
No, I have missed that. I will remove it.

uablrek replied:
Oh, I misunderstood, it _is_ removed, and the question was: should it?
And yes, modprobe is not necessary. Described in https://github.com/kubernetes/kubernetes/pull/114669#issuecomment-1363790134.
Other functions, like conntrack and iptables relies on auto-load already, so ipvs-proxier shoudn't need to do explicit modprobes.

uablrek commented:
I added this code for test;
```go
	klog.V(5).InfoS("Virtual Servers after adding dummy", "count", len(vservers))
	if len(vservers) == 0 {
		klog.InfoS("Dummy VS not created", "scheduler", scheduler)
		return fmt.Errorf("Ipvs not supported") // This is a BUG work-around
	}
	klog.V(5).InfoS("Dummy VS created", "vs", vs)

	if scheduler == "wlc" {
		klog.Fatalln("Terminate for test reasons")
	}

	if err := ipvs.DeleteVirtualServer(&vs); err != nil {
		klog.ErrorS(err, "Could not delete dummy VS")
		return err
	}
```
I started with the default scheduler "rr" and switched to "wlc", the expected sequence is:
1. Start normally, alter to scheduler=wlc which trigger a kube-proxy re-start
2. Create a dummy VS for scheduler=wlc
3. Hit the Fatalf() and leave the dummy VS un-deleted
4. On re-start a VS with scheduler=wlc already exist, i.e. the lingering dummy VS
5. Take the happy path and return without any further tests, leaving the dummy VS as-is
6. Kube-proxy start continues with scheduler=wlc

The question is: will the lingering dummy VS be deleted?

And the answer is: yes, it will.

This is a fairly slow process so the interim-state between re-starts can be captured;
```
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  198.51.100.0:20000 wlc
TCP  12.0.0.1:443 rr
  -> 192.168.1.1:6443             Masq    1      0          0         
```
Here the old kubernetes svc is still using "rr", and the dummy VS can be spotted. But after the second re-start it is corrected:
```
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  12.0.0.1:443 wlc
  -> 192.168.1.1:6443             Masq    1      0          0         
```
The kubernetes svc now uses "wlc", and the dummy VS is deleted.

Commit:
proxy/ipvs: Describe and handle a bug in moby/ipvs

Handle https://github.com/moby/ipvs/issues/27
A work-around was already in place, but a segv would occur
when the bug is fixed. That will not happen now.

thockin commented:
I believe you that it is deleted but I can't see WHY it will be.

After the log-fatal, it restarts.  It hits `ipvs.GetVirtualServers()` which returns non-nil (I presume) and `len(vservers) > 0`, right?  It would find the dummy with wlc and return - so who is deleting it?

uablrek commented:
The dummy VS is deleted in the `cleanLegacyService()` function;

https://github.com/kubernetes/kubernetes/blob/685d639cb5f04c9c388866efc8b39dec80e15856/pkg/proxy/ipvs/proxier.go#L2121

This function is called from the `syncProxyRules()` function which is the main sync function. It can be seen in the logs from the second re-start in the example above:
```
I1226 08:45:14.289822    1315 iptables.go:423] "Running" command="iptables-restore" arguments=[-w 5 -W 100000 --noflush --counters]
I1226 08:45:14.292385    1315 proxier.go:2099] "Delete service" virtualServer="198.51.100.0:20000/TCP"
I1226 08:45:14.292433    1315 proxier.go:1002] "syncProxyRules complete" elapsed="56.913032ms"
I1226 08:45:14.292448    1315 proxier.go:1026] "Syncing ipvs proxier rules"
...
```

thockin reviewed (commented):
Thanks!

/lgtm
/approve
